tools/cuds-prioritized-attribution/terraform/main.tf

```diff
@@ -0,0 +1,185 @@
+resource "google_service_account" "cud_service_account" {
+  account_id   = "cascada-user"
+  display_name = "CUD Correction Service Account"
+}
+
+# Create Custom IAM Role
+resource "google_project_iam_custom_role" "cud_iam_role" {
+  role_id = "CudCorrectionIAMRole"
+  title   = "IAM Custom Role for CUD Correction Service Account"
+  permissions = [
+    "bigquery.jobs.create",
+    "bigquery.tables.create",
+    "bigquery.tables.updateData",
+    "bigquery.tables.get",
+    "bigquery.tables.getData"
+  ]
+}
+
+# Grant custom role to service account
+resource "google_project_iam_binding" "custom_role_binding" {
+  members = ["serviceAccount:${google_service_account.cud_service_account.email}"]
+  role    = "projects/${var.project_id}/roles/${google_project_iam_custom_role.cud_iam_role.role_id}"
+}
+
+# Grant Cloud Composer Role to service account
+resource "google_project_iam_binding" "composer_binding" {
+  members = ["serviceAccount:${google_service_account.cud_service_account.email}"]
+  role    = "roles/composer.worker"
+}
+
+# Grant BigQuery Permissions for Service Account on Corrected Dataset
+resource "google_bigquery_dataset" "corrected_dataset" {
+  dataset_id = "${var.corrected_dataset_id}"
+  location   = "${var.billing_export_location}"
+
+  access {
+    role          = "WRITER"
+    user_by_email = "${google_service_account.cud_service_account.email}"
+  }
+}
+
+# Create bucket for Composer temporary file store
+resource "google_storage_bucket" "commitment_file_store" {
+  name     = "${var.project_id}-cud-correction-commitment-data"
+  location = "${var.billing_export_location}"
+}
+
+# Create Composer Environment
+resource "google_composer_environment" "env" {
+  provider = "google-beta"
+  name     = "cud-correction-env"
+  region   = "${var.region}"
+  depends_on = ["google_project_service.composerapi",
+    "google_project_iam_binding.custom_role_binding",
+    "google_project_iam_binding.composer_binding",
+    "google_storage_bucket.commitment_file_store"
+  ]
+
+  config {
+    node_config {
+      zone            = "${var.zone}"
+      service_account = "${google_service_account.cud_service_account.email}"
+    }
+    software_config {
+      python_version = 3
+      airflow_config_overrides = {
+        update-pypi = "requirements.txt"
+      }
+
+      env_variables = {
+        project_id                  = "${var.project_id}"
+        billing_export_table_name   = "${var.billing_export_table_path}"
+        corrected_dataset_id        = "${var.corrected_dataset_id}"
+        corrected_table_name        = "${var.corrected_table_name}"
+        commitments_table_name      = "${var.commitment_table_path}"
+        enable_cud_cost_attribution = "${var.enable_cud_cost_attribution}"
+        cud_cost_attribution_option = "${var.cud_cost_attribution_option}"
+      }
+    }
+  }
+}
+
+
+# Upload dependencies folder to DAG folder in Composer Bucket
+resource "google_storage_bucket_object" "dag_init_upload" {
```

Comment:
I've been told it is poor form to deploy software like this with terraform.
Would cloud build be a more appropriate tool staging all these files for this?

Reply:
Possibly, I used the same terraform recommendations from this [tool](https://github.com/GoogleCloudPlatform/professional-services/tree/master/examples/ml-audio-content-profiling), as recommended by morgante.

Reply:
Respectfully, disagree that this is a similar pattern to the other tool
In the deployment process for the other tool you zip up source trees.
Here you require a contributor to manage each source file (including pesky `__init__.py`'s) as a separate GCS object which is much more error prone.
Having a deployment process that runs `gsutil -m rsync -r ${local_dags_dir} gs://${composer_bucket}/dags would be much preferable.

PTAL at this simple [cloud build example we have for deploying dags](https://github.com/GoogleCloudPlatform/professional-services/blob/master/examples/cloud-composer-cicd/) and see if you can adapt it for your use in this asset.

Aside,
We're working on a more fully baked CI/CD process for composer ([deck](https://docs.google.com/presentation/d/1xy5VRMqmdUVJUxBN5hmnYaATgZrGdkT9Ql9UPmYEwG0/edit#slide=id.g3bec517c07_32_443), [source](https://github.com/jaketf/ci-cd-for-data-processing-workflow)) for composer but it is not quite ready yet.

Reply:
done